satellite/satellitedb/overlaycache.go

```diff
@@ -902,6 +902,48 @@ func (cache *overlaycache) UpdateUptime(ctx context.Context, nodeID storj.NodeID
 	return getNodeStats(dbNode), Error.Wrap(tx.Commit())
 }
 
+// AllPieceCounts returns a map of node IDs to piece counts from the db.
+func (cache *overlaycache) AllPieceCounts(ctx context.Context) (_ map[storj.NodeID]int, err error) {
+	defer mon.Task()(&ctx, err)
+
+	rows, err := cache.db.All_Node_Id_Node_PieceCount_By_PieceCount_Not_Number(ctx)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	pieceCounts := make(map[storj.NodeID]int)
+
+	var nodeID storj.NodeID
+	for _, row := range rows {
+		// NB: assumes node IDs in nodes table are always valid.
+		copy(nodeID[:], row.Id)
+		pieceCounts[nodeID] = int(row.PieceCount)
+	}
+	return pieceCounts, nil
+}
+
+func (cache *overlaycache) UpdatePieceCounts(ctx context.Context, pieceCounts map[storj.NodeID]int) (err error) {
+	defer mon.Task()(&ctx, err)
+
+	var nodeIDs storj.NodeIDList
+	for nodeID := range pieceCounts {
+		nodeIDs = append(nodeIDs, nodeID)
+	}
+
+	var (
+		sqlQuery string
+		args     []interface{}
+	)
+	// TODO: optimize this
+	updateSQL := "UPDATE nodes SET ( piece_count ) = ( %d ) WHERE id == ?;"
```

Comment:
This will fail on postgres due to: "pq: cannot insert multiple commands into a prepared statement".  You'll need to explicitly pass the `id` value.

Reply:
Ethan is right, and also this will also complain on postgres due to the lack of a `==` operator. In most (all?) SQL dialects you want a single `=` to check for equality.

Reply:
And, heh, I don't think these parentheses will work- I think postgres will interpret the first set as a multi-column update, which would be fine, but then the second set will just be plain expression-grouping, and then pg will complain because the source for a multi-column update has to be something that can have multiple columns, like a sub-select.

tl;dr this should look something like `UPDATE nodes SET piece_count = %d WHERE id = E'\\x%x';`

but i'm going to try to figure out a better way to combine all these updates, because it kinda looks like _all_ the nodes are supposed to go in here. Maybe a temp table filled via `COPY FROM`, and then join the `UPDATE` with that?

Reply:
OK I think inline `VALUES` will be best, since the temp table approach would add more db round-trips. It would look something like this:

```go
if len(pieceCounts) == 0 {
    return nil
}
sqlQuery := `UPDATE nodes SET piece_count = newvals.piece_count FROM (VALUES `
args := make([]interface{}, 0, len(pieceCounts) * 2)
for nodeID, pieceCount := range pieceCounts {
    sqlQuery += `(?, ?), `
    args = append(args, nodeID, pieceCount)
}
sqlQuery = sqlQuery[:len(sqlQuery)-2] // trim off the last comma+space
sqlQuery += `) newvals(nodeid, piece_count) WHERE nodes.id = newvals.nodeid`

_, err = cache.db.DB.ExecContext(ctx, cache.db.Rebind(sqlQuery), args...)
```

Of course `UPDATE ... FROM` won't work in sqlite3, so you'll have to switch on the database driver type 😢

Reply:
Thanks @thepaul! We only need to be using postgres right now anyways.

Reply:
Oh, ok, that's excellent. Most satellite code has to work with both because our tests use sqlite.

Reply:
Don't we need sqlite support for the storj-sim? Or are you saying this will get resolved in another PR?

I get this when I run storj-sim
```
satellite/0      1256FwPPzP 20:37:56.126 | ERROR	garbage collection	gc/service.go:107	error updating piece counts	{"error": "satellitedb: near \"FROM\": syntax error", "errorVerbose": "satellitedb: near \"FROM\": syntax error\n\tstorj.io/storj/satellite/satellitedb.(*overlaycache).UpdatePieceCounts:951\n\tstorj.io/storj/satellite/satellitedb.(*lockedOverlayCache).UpdatePieceCounts:940\n\tstorj.io/storj/satellite/gc.(*Service).Run.func1:105\n\tstorj.io/storj/internal/sync2.(*Cycle).Run:87\n\tstorj.io/storj/satellite/gc.(*Service).Run:84\n\tstorj.io/storj/satellite.(*Peer).Run.func12:706\n\tgolang.org/x/sync/errgroup.(*Group).Go.func1:57"}
                                         | storj.io/storj/satellite/gc.(*Service).Run.func1
                                         | 	/Users/ethan/workspace/github/storj/satellite/gc/service.go:107
                                         | storj.io/storj/internal/sync2.(*Cycle).Run
                                         | 	/Users/ethan/workspace/github/storj/internal/sync2/cycle.go:87
                                         | storj.io/storj/satellite/gc.(*Service).Run
                                         | 	/Users/ethan/workspace/github/storj/satellite/gc/service.go:84
                                         | storj.io/storj/satellite.(*Peer).Run.func12
                                         | 	/Users/ethan/workspace/github/storj/satellite/peer.go:706
                                         | golang.org/x/sync/errgroup.(*Group).Go.func1
                                         | 	/Users/ethan/go/pkg/mod/golang.org/x/sync@v0.0.0-20190423024810-112230192c58/errgroup/errgroup.go:57
```

Reply:
@ethanadams yeah, storj-sim won't be able to update the piece counts in the DB but I don't think that's a big deal. Assuming you even needed storj-sim to gc, it will still build the in-memory piece counts.

Reply:
We don't need storj-sim updating the piece counts, but we should still ensure that all database actions work on both sqlite and postgres.

Reply:
That's fair

Reply:
In other parts of the codebase we use `switch t := db.db.Driver().(type) {` for this